vaadin-custom-field-flow-demo/src/main/java/com/vaadin/flow/component/customfield/vaadincom/CustomFieldView.java

```diff
@@ -1,161 +1,174 @@
+
+        PhoneNumberField() {
+            setLabel("Phone number");
+            countryCode.setItems("+358", "+46", "+34");
+            countryCode.getStyle().set("width", "6em");
+            countryCode.setPlaceholder("Code");
+            areaCode.setPattern("[0-9]*");
+            areaCode.setPreventInvalidInput(true);
+            areaCode.setMaxLength(4);
+            areaCode.setPlaceholder("Area");
+            areaCode.getStyle().set("width", "5em");
+            subscriberCode.setPattern("[0-9]*");
+            subscriberCode.setPreventInvalidInput(true);
+            subscriberCode.setMaxLength(8);
+            subscriberCode.setPlaceholder("Subscriber");
+            HorizontalLayout horizontalLayout = new HorizontalLayout();
+            horizontalLayout.add(countryCode, areaCode, subscriberCode);
+            add(horizontalLayout);
         }
 
         @Override
-        protected BigDecimal generateModelValue() {
-            try {
-                Function<TextField, Integer> toNumber = t ->
-                   !t.isEmpty() ? Integer.valueOf(t.getValue()) : 0;
-                int i = toNumber.apply(integerPart);
-                int f = toNumber.apply(fractionalPart);
-                return new BigDecimal(i + "." + f);
-            } catch (NumberFormatException e) {
-                return BigDecimal.ZERO;
-            }
+        protected String generateModelValue() {
+            return countryCode.getValue() + " " + areaCode.getValue() + " "
+                    + subscriberCode.getValue();
         }
 
         @Override
-        protected void setPresentationValue(BigDecimal newPresentationValue) {
+        protected void setPresentationValue(String newPresentationValue) {
             if (newPresentationValue == null) {
-                integerPart.setValue(null);
-                fractionalPart.setValue(null);
-            } else {
-                String[] parts = newPresentationValue.toPlainString()
-                    .split("\\.");
-                integerPart.setValue(parts[0]);
-                fractionalPart.setValue(parts.length > 1 ? parts[1] : null);
+                areaCode.setValue(null);
+                subscriberCode.setValue(null);
             }
         }
-
     }
     // end-source-example
 
     // begin-source-example
-    // source-example-heading: Single element wrapping
-    public static class SingleElementWrapping extends CustomField<String> {
-        private final TextField wrappedField = new TextField();
-
-        SingleElementWrapping() {
-            setLabel("Name");
-            add(wrappedField);
-        }
+    // source-example-heading: Value change event
+    public static class ValueChangeEvent extends VerticalLayout {
+        private final Div display = new Div();
 
-        @Override
-        protected String generateModelValue() {
-            return wrappedField.getValue();
+        ValueChangeEvent() {
+            // The code is reused from the previous example.
```

Comment:
I would probably remove padding as it creates too much free space around the field.

Reply:
Done.